src/Raven.Server/Documents/Queries/StreamQueryResult.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Raven.Client.Documents.Operations.Counters;
+using Raven.Server.Documents.Handlers;
+using Raven.Server.Documents.Includes;
+using Raven.Server.Documents.Queries.Explanation;
+using Raven.Server.ServerWide;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries
+{
+    public abstract class StreamQueryResult<T> : QueryResultServerSide<T>
+    {
+        private readonly IStreamQueryResultWriter<T> _writer;
+        private readonly OperationCancelToken _token;
+        private bool _anyWrites;
+        private bool _anyExceptions;
+
+        protected StreamQueryResult(HttpResponse response, IStreamQueryResultWriter<T> writer, OperationCancelToken token)
+        {
+            if (response.HasStarted)
+                throw new InvalidOperationException("You cannot start streaming because response has already started.");
+
+            _writer = writer;
+            _token = token;
+        }
+
+        public IStreamQueryResultWriter<T> GetWriter()
+        {
+            return _writer;
+        }
+
+        public OperationCancelToken GetToken()
+        {
+            return _token;
+        }
+
+        public bool GetAnyWrites()
+        {
+            return _anyWrites;
+        }
```

Comment:
BTW the above can be just properties e.g. `protected IStreamQueryResultWriter<T> Writer => _writer`